compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -78,10 +78,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // While we don't allow *arbitrary* coercions here, we *do* allow
         // coercions from ! to `expected`.
         if ty.is_never() {
-            assert!(
-                !self.typeck_results.borrow().adjustments().contains_key(expr.hir_id),
-                "expression with never type wound up being adjusted"
-            );
+            if self.typeck_results.borrow().adjustments().contains_key(expr.hir_id) {
+                self.tcx().sess.delay_span_bug(
+                    expr.span,
+                    "expression with never type wound up being adjusted",
+                );
+                return self.tcx().ty_error();
+            }
+
```

Comment:
It would be nice to return the adjusted type from here instead of `ty_error`.

```suggestion
            if let Some(adjustments) = self.typeck_results.borrow().adjustments().get(expr.hir_id) {
                self.tcx().sess.delay_span_bug(
                    expr.span,
                    "expression with never type wound up being adjusted",
                );
                if let [Adjustment { kind: Adjust::NeverToAny, target }] = &adjustments[..] {
                    return target;
                } else {
                    return self.tcx().ty_error();
                }
            }

```

Reply:
That's a great idea! Thank you!